catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CategoryListFragment.java

```diff
@@ -238,47 +243,115 @@ public void onItemClick(CategoryListItem item) {
-						.addResourceToActiveFormula(item.nameResId);
-				getActivity().onBackPressed();
+				if (LIST_FUNCTIONS.contains(item.nameResId)) {
+					onUserListFunctionSelected(item);
+				} else {
+					addResourceToActiveFormulaInFormulaEditor(item);
+					getActivity().onBackPressed();
+				}
 				break;
 		}
 	}
 
+	private FormulaEditorFragment addResourceToActiveFormulaInFormulaEditor(CategoryListItem categoryListItem) {
+		FormulaEditorFragment formulaEditorFragment = ((FormulaEditorFragment) getFragmentManager()
+				.findFragmentByTag(FormulaEditorFragment.FORMULA_EDITOR_FRAGMENT_TAG));
+		formulaEditorFragment.addResourceToActiveFormula(categoryListItem.nameResId);
+		return formulaEditorFragment;
+	}
+
+	private void addResourceToActiveFormulaInFormulaEditor(CategoryListItem categoryListItem, UserList lastUserList) {
+		addResourceToActiveFormulaInFormulaEditor(categoryListItem).addUserListToActiveFormula(lastUserList.getName());
+		getActivity().onBackPressed();
+	}
+
+	private void onUserListFunctionSelected(CategoryListItem item) {
+		FragmentActivity activity = getActivity();
+		TextInputDialog.Builder builder = new TextInputDialog.Builder(activity);
+		final List<UserList> projectUserList =
+				ProjectManager.getInstance().getCurrentProject().getUserLists();
+		final List<UserList> spriteUserList =
+				ProjectManager.getInstance().getCurrentSprite().getUserLists();
+		insertLastUserListToActiveFormula(item, projectUserList, spriteUserList,
+				activity, builder);
+	}
+
+	@VisibleForTesting
+	public void insertLastUserListToActiveFormula(CategoryListItem categoryListItem,
+			List<UserList> projectUserList, List<UserList> spriteUserList,
+			FragmentActivity activity,
+			TextInputDialog.Builder builder) {
+
+		if (spriteUserList.isEmpty() && projectUserList.isEmpty()) {
+			showNewUserListDialog(categoryListItem, projectUserList, spriteUserList,
+					activity, builder);
+			return;
+		}
+
+		if (!spriteUserList.isEmpty()) {
+			addResourceToActiveFormulaInFormulaEditor(categoryListItem, spriteUserList.get(spriteUserList.size() - 1));
+			return;
+		}
+		addResourceToActiveFormulaInFormulaEditor(categoryListItem, projectUserList.get(projectUserList.size() - 1));
+	}
+
+	private void showNewUserListDialog(CategoryListItem categoryListItem, List<UserList> projectUserList, List<UserList> spriteUserList,
+			FragmentActivity activity, TextInputDialog.Builder builder) {
+		builder.setHint(activity.getString(R.string.data_label))
+				.setPositiveButton(activity.getString(R.string.ok), (TextInputDialog.OnClickListener) (dialog, textInput) -> {
+					if (textInput.isEmpty()) {
+						activity.onBackPressed();
+						return;
+					}
+					UserList userList = new UserList(textInput);
+
+					RadioButton addToProjectListsRadioButton = ((Dialog) dialog).findViewById(R.id.global);
+					boolean addToProjectLists = addToProjectListsRadioButton.isChecked();
+
+					if (addToProjectLists) {
```

Comment:
Please make a util method/class, or something else for this task. Because there are already two places where the user-list is added to the project.